### PR TITLE
[ROCm][gfx11x] Pad W4A16 dequant-prefill weights off the 2K cache cliff

### PR DESCRIPTION
**Pull request by roberteg16**
## Summary

On gfx11x (RDNA3, e.g. Strix Halo / gfx1151) a dense weight whose row stride lands on a 2048-byte multiple makes the multi-row global loads collide on the L2/MALL channel hash, so hipBLASLt and the stride-aware wvSplitK kernel stall on a bandwidth cliff. The fix is to offset the row stride by one cache line (+128 B), which keeps each row cache-line aligned while moving the stride off the collision grid. This PR brings that mitigation to the W4A16 prefill-dequant path and unifies the cliff threshold at 2048 bytes across the codebase.

## Motivation

The int4 skinny path already dodges this cliff (it pads when the packed row stride is a 2048-byte multiple), but two gaps remained. First, `UnquantizedLinearMethod` guarded the coarser 4096-byte stride only, so dense weights sitting on a 2048-but-not-4096 byte stride were left on the cliff. Second, the W4A16 prefill-dequant copy (`VLLM_W4A16_PREFILL_DEQUANT`) ran a raw `torch.nn.functional.linear` with no stride mitigation at all, so dequantizing a quantized layer for prefill could actually be *slower* than the int4 path for the affected shapes.

## Changes

This PR is split into two commits. The first lowers the `UnquantizedLinearMethod` cliff-pad threshold from a 4096-byte row stride to 2048 bytes, matching the int4 skinny path and reflecting that the channel-hash collision is a 2048-byte phenomenon. The second extracts the pad into a shared `pad_weights_avoid_cache_cliff_on_gfx11` helper (a no-op off gfx11x), applies it to the cached W4A16 dequant-prefill weight, accounts for the extra cache line in the `gpu_memory_utilization` budget gate, and routes the dequant matmul through `rocm_unquantized_gemm_impl` so the dense dequant path reuses the same stride-aware GEMM selection as the unquantized linear path.

## Results

Measured on gfx1151 (Strix Halo), Qwen3.6-35B-A3B-W4A16, prefill at M=994 (input-len 100, single multimodal request), comparing the int4 path, the dequant path before padding, and the dequant path with this PR. Times are total GPU time over the run; TF is achieved TFLOPS; % is peak MFMA utilization.

| Shape (M×N×K) | int4 (no dequant) | dequant, no pad | dequant + 2K pad (this PR) |
|---|---|---|---|
| 994×12288×2048 (×30) | 62.02 ms · 24.2 TF · 56% | 47.83 ms · 31.4 TF · 73% | 41.34 ms · 36.3 TF · 84% |
| 994×9216×2048 (×10) | 15.33 ms · 24.5 TF · 57% | 11.78 ms · 31.9 TF · 74% | 9.87 ms · 38.0 TF · 89% |
| 994×2048×4096 (×40) | 26.40 ms · 25.3 TF · 59% | 29.51 ms · 22.7 TF · 53% | 19.23 ms · 34.8 TF · 81% |
| Dense prefill subtotal | 103.75 ms | 89.12 ms | 70.44 ms |
| TTFT | 621 ms | 599 ms | 581 ms |

The 994×2048×4096 shape is the one that regressed under unpadded dequant (it was the only shape the int4 path already cache-padded); padding the dequant copy recovers it and beats the int4 baseline by 27%. **The two wider shapes gain because the 2048-byte threshold catches strides the old 4096-byte rule missed.** MoE experts, the vision encoder, and the lm_head are unchanged, confirming the change is scoped to the dense prefill GEMMs.

### CODE REVIEW

File: vllm/model_executor/kernels/linear/mixed_precision/hybrid_w4a16.py [hunk 1]
```diff
@@ -896,7 +901,20 @@ def _maybe_cache_dequant_prefill_weight(
 
         c = self.config
         N, K = unpacked.shape
-        dequant_bytes = N * K * torch.empty((), dtype=c.act_type).element_size()
+        elem_size = torch.empty((), dtype=c.act_type).element_size()
+        # Account for the gfx11x cache-line pad applied to the cached copy below
+        # (pad_weights_avoid_cache_cliff_on_gfx11) so the gpu_memory_utilization gate
+        # stays honest.
+        pad_elems = (
```
Comment (mgehre-amd) [inline vllm/model_executor/kernels/linear/mixed_precision/hybrid_w4a16.py:910]: I think you can extract that into a function and share it with pad_weights_avoid_cache_cliff_on_gfx11

**Comment by mgehre-amd** [comment]
@vipangul, could you please check whether this PR helps Qwen3-VL-4B on Hawk point when VLLM_W4A16_PREFILL_DEQUANT=1?

File: vllm/model_executor/kernels/linear/mixed_precision/hybrid_w4a16.py (hunk 1 above)
Comment (roberteg16) [reply vllm/model_executor/kernels/linear/mixed_precision/hybrid_w4a16.py:910]: Just was doing that 🙈

**Review by mgehre-amd — COMMENTED** [review]
Looks good!

**Review by mgehre-amd — APPROVED** [review]
